src/mapgen_carpathian.cpp

```diff
@@ -0,0 +1,458 @@
+
+	// Generate the registered decorations
+	if (flags & MG_DECORATIONS)
+		m_emerge->decomgr->placeAllDecos(this, blockseed, node_min, node_max);
+
+	// Generate the registered ores
+	m_emerge->oremgr->placeAllOres(this, blockseed, node_min, node_max);
+
+	// Sprinkle some dust on top after everything else was generated
+	dustTopNodes();
+
+	// Update liquids
+	updateLiquid(&data->transforming_liquid, full_node_min, full_node_max);
+
+	// Calculate lighting
+	if (flags & MG_LIGHT) {
+		calcLighting(node_min - v3s16(0, 1, 0), node_max + v3s16(0, 1, 0),
+				full_node_min, full_node_max);
+	}
+
+	this->generating = false;
+}
+
+
+///////////////////////////////////////////////////////////////////////////////
+
+
+int MapgenCarpathian::getSpawnLevelAtPoint(v2s16 p)
+{
+	s16 level_at_point = terrainLevelAtPoint(p.X, p.Y);
+	if (level_at_point <= water_level || level_at_point > water_level + 32)
+		return MAX_MAP_GENERATION_LIMIT; // Unsuitable spawn point
+	else
+		return level_at_point;
+}
+
+
+float MapgenCarpathian::terrainLevelAtPoint(s16 x, s16 z)
+{
+	float ground = NoisePerlin2D(&noise_base->np, x, z, seed);
+	float height1 = NoisePerlin2D(&noise_height1->np, x, z, seed);
+	float height2 = NoisePerlin2D(&noise_height2->np, x, z, seed);
+	float height3 = NoisePerlin2D(&noise_height3->np, x, z, seed);
+	float height4 = NoisePerlin2D(&noise_height4->np, x, z, seed);
+	float hter = NoisePerlin2D(&noise_hills_terrain->np, x, z, seed);
+	float rter = NoisePerlin2D(&noise_ridge_terrain->np, x, z, seed);
+	float ster = NoisePerlin2D(&noise_step_terrain->np, x, z, seed);
+	float n_hills = NoisePerlin2D(&noise_hills->np, x, z, seed);
+	float n_ridge_mnt = NoisePerlin2D(&noise_ridge_mnt->np, x, z, seed);
+	float n_step_mnt = NoisePerlin2D(&noise_step_mnt->np, x, z, seed);
+
+	int height = -MAX_MAP_GENERATION_LIMIT;
+
+	for (s16 y = 1; y <= 30; y++) {
+		float mnt_var = NoisePerlin3D(&noise_mnt_var->np, x, y, z, seed);
+
+		// Gradient
+		s32 grad = terrain_zero_level - y;
+		// Shallow seabed
+		if (y < water_level) {
+			grad = grad_wl + (water_level - y) * 3;
+		}
+
+		// Hill/Mountain height (hilliness)
+		float hill1 = getLerp(height1, height2, mnt_var);
+		float hill2 = getLerp(height3, height4, mnt_var);
+		float hill3 = getLerp(height3, height2, mnt_var);
+		float hill4 = getLerp(height1, height4, mnt_var);
+		float hilliness = std::fmax(std::fmin(hill1, hill2), std::fmin(hill3, hill4));
+
+		// Rolling hills
+		float hill_mnt = hilliness * pow(n_hills, 2.f);
+		float hills = pow(hter, 3.f) * hill_mnt;
+
+		// Ridged mountains
+		float ridge_mnt = hilliness * (1.f - fabs(n_ridge_mnt));
+		float ridged_mountains = pow(rter, 3.f) * ridge_mnt;
+
+		// Step (terraced) mountains
+		float step_mnt = hilliness * getSteps(n_step_mnt, fabs(mnt_var));
+		float step_mountains = pow(ster, 3.f) * step_mnt;
+
+		// Final terrain level
+		float mountains = hills + ridged_mountains + step_mountains;
+		float density = ground + mountains + grad;
```

Comment:
Please rename 'density' to 'surface level' or 'surface y' to avoid confusion with 'density mehtod' mapgens which this is not.